### PR TITLE
[Table] Correct pureRender logic on Cell as style object is reconstructed each render

### PR DESCRIPTION
**Pull request by zerovox**
@dcervelli This fixes the cell re-renders when only the style object has changed. 

Here's an after/before on this change in the profiler.

screen shot 2017-05-16 at 16 22 18

I have a commit locally that applies the same treatment to RowHeaderCell and ColumnHeaderCell, however it doesn't quite do the right thing. Both take a prop `resizeHandle` which is a React Component. This component, as with all components, is recreated each time, and as such it fails at a shallow equality check. I wonder if there's a react utility for checking ahead of time wether two elements will work out to be equal?

### CODE REVIEW

**Comment by zerovox** [comment]
In fact, both `RowHeader` and `ColumnHeader` create new callback functions in their `renderCell` methods, causing `Resizable` to get re-rendered causing, each `HeaderCell` and `RowCell` to get re-rendered. The change proposed above causes `Cell` to correctly pure render, if `RowHeader` and `ColumnHeader` were optimized w.r.t these bound callbacks in such a way that `Resizable` was only re-rendering when non-callback props change, I estimate you could get somewhere between a 1.5x and 2x frame rate on the table. I don't really see anywhere with the current table architecture to go beyond that.

**Comment by cmslewis** [comment]
Thanks for the work, @zerovox! Can you correct the failing tests and lint checks? Then I'll take another look.

File: packages/table/src/cell/cell.tsx [hunk 1]
```diff
@@ -53,17 +53,35 @@ export interface ICellProps extends IIntentProps, IProps {
...
-@PureRender
 export class Cell extends React.Component<ICellProps, {}> {
     public static defaultProps = {
         truncated: true,
         wrapText: false,
     };
 
+    public shouldComponentUpdate(nextProps: ICellProps) {
+        // shallowly comparable props like "className" tend not to change in the default table
+        // implementation, so do that check last with hope that we return earlier and avoid it
```
Comment (giladgray) [inline packages/table/src/cell/cell.tsx]: do the shallow check _first_?

Comment (cmslewis) [inline packages/table/src/cell/cell.tsx]: I believe this text is just copied-and-pasted from a similar series of checks we do in another component. Makes sense to do the shallow stuff first I guess. (Original motivation was to first check the stuff most likely to have changed.)

Comment (adidahiya) [inline packages/table/src/cell/cell.tsx]: Yeah I think you should check the props that are mostly likely to change first

**Comment by cmslewis** [comment]
@zerovox @dcervelli I'll see if I can get these tests and lint issues fixed for you (though LMK if you're already close).

File: packages/table/src/headers/rowHeaderCell.tsx [hunk 2]
```diff
@@ -67,13 +67,34 @@ export interface IRowHeaderState {
...
-@PureRender
 export class RowHeaderCell extends React.Component<IRowHeaderCellProps, IRowHeaderState> {
     public state = {
         isActive: false,
     };
 
+    public shouldComponentUpdate(nextProps: IRowHeaderCellProps) {
+        // shallowly comparable props like "className" tend not to change in the default table
+        // implementation, so do that check last with hope that we return earlier and avoid it
+        // altogether.
```
Comment (cmslewis) [inline packages/table/src/headers/rowHeaderCell.tsx]: Can nix this comment.

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 3]
```diff
@@ -135,6 +151,14 @@ export class ColumnHeaderCell extends React.Component<IColumnHeaderCellProps, IC
         isActive: false,
     };
 
+    public shouldComponentUpdate(nextProps: IColumnHeaderCellProps) {
+        // shallowly comparable props like "className" tend not to change in the default table
+        // implementation, so do that check last with hope that we return earlier and avoid it
+        // altogether.
```
Comment (cmslewis) [inline packages/table/src/headers/columnHeaderCell.tsx]: Can nix this comment.

**Comment by cmslewis** [comment]
_// Putting on my cowboy hat :cowboy_hat_face:_

@zerovox: @giladgray and I chatted; FYI I'm going to merge this now that all the tests are passing, let `master` be broken for a moment, then open a PR to fix the lint issues. Here we go!

**Comment by llorca** [comment]
👏👏👏

Thanks @zerovox 😍

**Pull request merged**